### PR TITLE
Completing milestone

### PR DESCRIPTION
**Pull request by nlukic97**
- Login POST method
- Register POST method
- data validation and sanitization for registration and login
- JWT token implemented
- Authentication middleware implemented in `/getAllUsers` route
- manual type declarations added to the User model for typescript compatibility

### CODE REVIEW

File: ts/controllers/User.controller.ts [hunk 1]
```diff
@@ -1,45 +1,135 @@
...
+import jwt from 'jsonwebtoken'
 
 interface ReqRes {
   (req: Request, res: Response): unknown
 }
 
 // methods
+/* --- Getting the number of users */
 const getAllUsers: ReqRes = async (req, res) => {
   const users = await User.findAll()
```
Comment (vuxor) [inline ts/controllers/User.controller.ts:15]: This is not optimized query. There is a separate query only for count

File: ts/controllers/User.controller.ts [hunk 2]
```diff
@@ -1,45 +1,135 @@
...
-const insertNewUser: ReqRes = async (req, res) => {
-  const user = await User.build({
-    username: req.body.username,
-    email: req.body.email,
-    password: req.body.password,
-  })
+/* --- User registration --- */
+const register: ReqRes = async (req, res) => {
+  const keys = Object.keys(req.body)
+  if(!keys.includes('username') || !keys.includes('email') || !keys.includes('password')){
```
Comment (vuxor) [inline ts/controllers/User.controller.ts:24]: If you only check for keys, values can still be empty

File: ts/controllers/User.controller.ts [hunk 3]
```diff
@@ -1,45 +1,135 @@
...
+        {username:data.username},
+      ]
+    }
+    })
 
+  if(userExists){
+    const match = [];
+    
+    if(userExists.email === data.email) match.push('email')
+    if(userExists.username === data.username) match.push('userame')
```
Comment (vuxor) [inline ts/controllers/User.controller.ts:56]: typo

File: ts/controllers/User.controller.ts [hunk 4]
```diff
@@ -1,45 +1,135 @@
...
+      res.status(403).json(errorMessages)
     })
 }
 
+/* --- User login --- */
+const login: ReqRes = async (req, res)=>{
+
+  // Making sure the 
+  const keys:Array<string|null> = Object.keys(req.body)
+  if(!keys.includes('name') || !keys.includes('password')){
```
Comment (vuxor) [inline ts/controllers/User.controller.ts:96]: You should check for values also

**Review by vuxor — CHANGES_REQUESTED** [review]

File: ts/controllers/User.controller.ts (hunk 1 above)
Comment (nlukic97) [reply ts/controllers/User.controller.ts]: Optimized the query to `count`. I've also updated the route and method from `getAllusers` to `countAllusers`.

File: ts/controllers/User.controller.ts (hunk 3 above)
Comment (nlukic97) [reply ts/controllers/User.controller.ts]: Woops, missed that one. Fixed it, thank you.

File: ts/controllers/User.controller.ts (hunk 2 above)
Comment (nlukic97) [reply ts/controllers/User.controller.ts]: I've updated this to specifically check all submitted keys are non-empty strings before checking anything else:

https://github.com/nlukic97/uteam-api/commit/4ea7a3a5516074fd371619f198293eda2fb37dbd#diff-78cee36d0b36a1c2bbab287cd94af0fb80fdf7568713df0b1e7c9600ea350303R23-R36

File: ts/controllers/User.controller.ts (hunk 4 above)
Comment (nlukic97) [reply ts/controllers/User.controller.ts]: Changed the method to resemble the register method by first checking all required body params are there, and that they are non-empty strings. After this, specific data validation is performed.

https://github.com/nlukic97/uteam-api/commit/4ea7a3a5516074fd371619f198293eda2fb37dbd#diff-78cee36d0b36a1c2bbab287cd94af0fb80fdf7568713df0b1e7c9600ea350303R121-R133

**Comment by nlukic97** [comment]
Hey, I've fixed all the mistakes + added the required **regex check for usernames**: https://github.com/nlukic97/uteam-api/blob/ab30e5bc2398e2a9bd8655f629afbe159d92cfc5/ts/controllers/User.controller.ts#L59-L62

**Pull request merged**